### ISSUE TITLE
@protocol functions with no parameters are not found by robovm bro

### ISSUE THREAD

**Issue by alexitelman** [robovm/robovm#1053]
When trying to bind a protocol function with no parameters I'm getting the 'no selector found' error.
If I add 2 parameters it's bound successfully.
Even with just one parameter it gives an error.

For example:

@protocol MyProto
  -(void)myMethod;
@end

Binding:

interface MyProto extends NSObjectProtocol {
    @Method(selector = "myMethod")
    void myMethod();
}

Error:
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[MyImpl myMethod]: unrecognized selector sent to instance 0x17570870'

**Comment by ntherning** [comment]
What does the implementation look like? How do you use this binding from Java? Please post some more code to give us some more context. A complete reproduction case would be most useful.

**Comment by alexitelman** [comment]
Actually the protocol is in Swift, but this also happened in obj-c.
The protocol looks like this:
@objc
public protocol MyProto {
    func myMethod(  )
}

The impl looks like this:
public class MyImpl : MyProto {
    func myMethod( ) {
        // do stuff
    }
}

I have a method 'setMyProto' in another protocol 'MyApi' :

@objc
public protocol MyApi {
    func setMyProto( myProto: MyProto )
}

And another static class: 

@objc
public class MyStatic : NSObject {
   static var mMyApi: MyApi!
   public static func setMyApi( myApi: MyApi ) {
       mMyApi = myApi
   }
}

In Java I wrap the MyStatic class like this:

@NativeClass("MyFramework.MyStatic")
public class MyStatic extends NSObject {
    @Method(selector = "setMyApi:")
    private static native void nativeSetMyApi(MyApi myApi);

    public static void setMyApi(MyApi myApi){
        nativeSetMyApi(myApi);
    }
}

public interface MyApi extends NSObjectProtocol {
    @Method
    void setMyProto(MyProto myProto);
}

public interface MyProto extends NSObjectProtocol {
    @Method(selector = "myMethod")
    void myMethod();
}

public class MyApiImpl extends NSObject implements MyApi {
    private MyProto mMyProto;
    @Override
    public void setMyProto(MyProto myProto) {
          mMyProto = myProto;
    }

   public void doStuff() {
         // other code...
         mMyProto.myMethod();
   }
}

in Main.java I set:

MyApi api = new MyApiImpl();
MyStatic.setMyApi(api);

and in some point in the code the doStuff method is invoked:

api.doStuff();



Actually, now I see that in the MyApi protocol I have methods with no parameters that are working, the difference is that I don't specify the selector there.

**Comment by alexitelman** [comment]
Yes, it seems that if I remove the selector in the @Method annotation then it works... strange.

**Comment by ntherning** [comment]
I'm not up to speed with Swift but all examples at https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Protocols.html appears to have the `@objc` annotation on both the protocol and the implementing Swift class. Maybe you need this too?

**Comment by alexitelman** [comment]
I tried that, it didn't help.
My impl is actually extending UIViewController, maybe it has something to
do with that?

**Comment by ntherning** [comment]
Please create a test app for us that recreates the issue and we'll look into it.

**Comment by alexitelman** [comment]
Here you go:
https://drive.google.com/file/d/0BwxZsZfwfINWVjBIM2tsQXlRZWM/view

This is the crash:

2015-07-06 18:48:32.164 Main[30955:2139516] -[ProtocolsIssueFramework.MyProtoImpl myMethod]: unrecognized selector sent to instance 0x7fb91b634e40
2015-07-06 18:48:32.185 Main[30955:2139516] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[ProtocolsIssueFramework.MyProtoImpl myMethod]: unrecognized selector sent to instance 0x7fb91b634e40'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000110ee6c65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x00000001105c9bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x0000000110eee0ad -[NSObject(NSObject) doesNotRecognizeSelector:] + 205
	3   CoreFoundation                      0x0000000110e43f4f ___forwarding___ + 495
	4   CoreFoundation                      0x0000000110e43cd8 _CF_forwarding_prep_0 + 120
	5   Main                                0x000000010d146f1e [J]org.robovm.objc.$M.void_objc_msgSendSuper(Lorg/robovm/objc/ObjCSuper;Lorg/robovm/objc/Selector;)V + 151
	6   Main                                0x000000010d149e08 [J]org.robovm.objc.$M.void_objc_msgSend_instance(Lorg/robovm/apple/foundation/NSObject;Lorg/robovm/objc/Selector;)V + 63
	7   Main                                0x000000010d14572f [j]org.robovm.objc.$M.void_objc_msgSend_instance(Lorg/robovm/apple/foundation/NSObject;Lorg/robovm/objc/Selector;)V[clinit] + 61
	8   Main                                0x000000010cc34052 [J]com.mycompany.myapp.MyProto$ObjCProxy.myMethod()V + 34
	9   Main                                0x000000010cc33e62 [J]com.mycompany.myapp.MyApiImpl.callProtoMethod()V + 23
	10  Main                                0x000000010cc34a58 [J]com.mycompany.myapp.MyViewController.clicked()V + 157
	11  Main                                0x000000010cc34ace [J]com.mycompany.myapp.MyViewController.$cb$clicked(Lcom/mycompany/myapp/MyViewController;Lorg/robovm/objc/Selector;)V + 19
	12  Main                                0x000000010cc348f5 [j]com.mycompany.myapp.MyViewController.$cb$clicked(Lcom/mycompany/myapp/MyViewController;Lorg/robovm/objc/Selector;)V[callback] + 146
	13  UIKit                               0x000000010ed98da2 -[UIApplication sendAction:to:from:forEvent:] + 75
	14  UIKit                               0x000000010eeaa54a -[UIControl _sendActionsForEvents:withEvent:] + 467
	15  UIKit                               0x000000010eea9919 -[UIControl touchesEnded:withEvent:] + 522
	16  UIKit                               0x000000010ede5998 -[UIWindow _sendTouchesForEvent:] + 735
	17  UIKit                               0x000000010ede62c2 -[UIWindow sendEvent:] + 682
	18  UIKit                               0x000000010edac581 -[UIApplication sendEvent:] + 246
	19  UIKit                               0x000000010edb9d1c _UIApplicationHandleEventFromQueueEvent + 18265
	20  UIKit                               0x000000010ed945dc _UIApplicationHandleEventQueue + 2066
	21  CoreFoundation                      0x0000000110e1a431 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	22  CoreFoundation                      0x0000000110e102fd __CFRunLoopDoSources0 + 269
	23  CoreFoundation                      0x0000000110e0f934 __CFRunLoopRun + 868
	24  CoreFoundation                      0x0000000110e0f366 CFRunLoopRunSpecific + 470
	25  GraphicsServices                    0x0000000115325a3e GSEventRunModal + 161
	26  UIKit                               0x000000010ed97900 UIApplicationMain + 1282
	27  Main                                0x000000010d02a6f0 [J]org.robovm.apple.uikit.UIApplication.main(ILorg/robovm/rt/bro/ptr/BytePtr$BytePtrPtr;Ljava/lang/String;Ljava/lang/String;)I + 195
	28  Main                                0x000000010d029bfb [J]org.robovm.apple.uikit.UIApplication.main([Ljava/lang/String;Ljava/lang/Class;Ljava/lang/Class;)V + 944
	29  Main                                0x000000010d028ea0 [j]org.robovm.apple.uikit.UIApplication.main([Ljava/lang/String;Ljava/lang/Class;Ljava/lang/Class;)V[clinit] + 68
	30  Main                                0x000000010cc33567 [J]com.mycompany.myapp.Main.main([Ljava/lang/String;)V + 220
	31  Main                                0x000000010d19332e _call0 + 142
	32  Main                                0x000000010d188450 callVoidMethod + 104
	33  Main                                0x000000010d18ac46 rvmCallVoidClassMethodA + 277
	34  Main                                0x000000010d18ad1a rvmCallVoidClassMethod + 134
	35  Main                                0x000000010d182da0 rvmRun + 511
	36  Main                                0x000000010d178af0 main + 301
	37  libdyld.dylib                       0x0000000110279145 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException

Process finished with exit code 0

**Comment by alexitelman** [comment]
Any progress on this?

**Comment by badlogic** [comment]
I'm afraid not. We are currently working to get the 1.5 release out.

**Comment by alexitelman** [comment]
Hello guys,
is this going to be fixed any time soon?
It makes our code very ugly

**Comment by alexitelman** [comment]
what more info do you need?

**Comment by badlogic** [comment]
None, we just need to get to the issue. Which we won't do on the weekend :)

**Comment by alexitelman** [comment]
Any time soon maybe guys?

**Comment by alexitelman** [comment]
Guys are you going to fix this or should I stop waiting?

**Comment by badlogic** [comment]
I'll start fixing this tomorrow.

**Comment by alexitelman** [comment]
How's it going? Need help?

**Comment by badlogic** [comment]
No progress so far i'm afraid. I'll let you know here.

**Comment by badlogic** [comment]
I recreated the Swift test case using ObjC, slightly simplified:

```objc
@interface ObjCApi : NSObject
+(id<ObjCProto>)getProtocol;
@end

@implementation ObjCApi
+(id<ObjCProto>)getProtocol {
    id<ObjCProto> proto = [[ObjCProtoImpl alloc] init];
    return proto;
}

@protocol ObjCProto
-(void)myMethod;
@end

@interface ObjCProtoImpl : NSObject <ObjCProto>
-(void)myMethod;
@end

@implementation ObjCProtoImpl: NSObject
-(void)myMethod {
    NSLog(@"Hello World");
}
@end
```

```java
@NativeClass("ObjCApi")
@Library(Library.INTERNAL)
public class ObjCApi extends NSObject {
    @Method(selector="getProtocol")
    public static native ObjCProto getProtocol();
}

public interface ObjCProto extends NSObjectProtocol {
    @Method(selector="myMethod")
    void myMethod();
}

@NativeClass("ObjCProtoImpl")
@Library(Library.INTERNAL)
public class ObjCProtoImpl extends NSObject implements ObjCProto {
    @Method(selector="myMethod")
    @Override
    public native void myMethod();
}
```

The following test app works as intended:

```java
        ObjCProto objcProto = new ObjCProtoImpl();
        objcProto.myMethod();
        
        ObjCProto objCProto2 = ObjCApi.getProtocol();
        objCProto2.myMethod();
```

I recreated the exact same setup with Swift, and calling the method `myMethod` fails as reported by @alexitelman. I have a feeling the problem lies with the namespacing Swift does for classes/protocols annotated with `@objc`.

**Comment by badlogic** [comment]
I reworked the reproduction sample, and i used RoboVM 1.7.1 (your Gradle build was set to 1.4.1-SNAPSHOT). I got the Swift version to work, here's the Swift code (single file):

```swift
import Foundation

@objc
public protocol SwiftProto {
    func myMethod()
}

@objc
public class SwiftProtoImpl : NSObject, SwiftProto {
    public func myMethod() {
        NSLog("In myMethod ...")
    }
}

@objc
public class OtherSwiftProtocol : NSObject, SwiftProto {
    public func myMethod() {
        NSLog("In myMethod ...")
    }
}

@objc
public class SwiftApi : NSObject {
    public static func getProtocol() -> SwiftProto {
        return SwiftProtoImpl()
    }
    
    public static func getOtherProtocol() -> OtherSwiftProtocol {
        return OtherSwiftProtocol()
    }
}
```

Java bindings

``java
@NativeClass("ProtocolsIssueFramework.SwiftApi")
public class SwiftApi extends NSObject {
    @Method(selector="getProtocol")
    public static native SwiftProto getProtocol();
    
    @Method(selector="getOtherProtocol")
    public static native SwiftProto getOtherProtocol();
}

public interface SwiftProto extends NSObjectProtocol {
    @Method(selector="myMethod")
    void myMethod();
}

@NativeClass("ProtocolsIssueFramework.SwiftProtoImpl")
public class SwiftProtoImpl extends NSObject implements SwiftProto {
    @Method(selector="myMethod")
    @Override
    public native void myMethod();
}
```

Here's the output:
![http://sht.tl/QU9vbz](http://sht.tl/QU9vbz)

Note how we get a `SwiftProtoImpl` in the first two cases, because we have a Java-side class for that. For the third getter, we get a `SwiftProto$ObjCProxy`, as we don't have a Java class for `SwiftOtherProtoImpl`. This is were your original sample failed. I'm now trying to fix this in your original sample as well.

**Comment by alexitelman** [comment]
Mario, this is not a Swift issue, this happens exactly the same in obj-c. Do you need a sample project for that too?

**Comment by badlogic** [comment]
No, i'll work with the sample project you submitted. Swift has one complication in that it namespaces classes and protocols. I suspect that the ObjC bridge doesn't handle this properly. I'll report back once i fixed up your sample project or fixed the potential issue in the bridge.

**Comment by badlogic** [comment]
I just recompiled your framework with Xcode 7, and recompiled the Java side, seems to work as expected with RoboVM 1.7.0:

![http://sht.tl/qt9oAh](http://sht.tl/qt9oAh)

The only thing i did was move the two lines initializing MyStatic from Main to the view controller.

Which RoboVM version are you on? Which Xcode version do you use? I'll try with Xcode 6.4 next.

**Comment by alexitelman** [comment]
I'm using 1.7 and 7 but I haven't tried this for a while, I'll try now.

**Comment by badlogic** [comment]
Ok, i just tried with 1.7 and 6.4 and it works as well ![http://sht.tl/EH1qUv](http://sht.tl/EH1qUv)

**Comment by alexitelman** [comment]
Seems to be working in a small test project, will test now in our large project.

**Comment by alexitelman** [comment]
Seems ok in our project too now. 
Great!
I wonder what made it work... did you fix something related in 1.7?

**Comment by badlogic** [comment]
We fixed a lot of things, but i couldn't pinpoint which change exactly fixed it. I think we have a binding related fix in the RT in 1.5 or 1.6.

Great that it works now, thanks for reporting!
